utils/sysfs/sysfs.go

```diff
@@ -335,25 +341,143 @@ func (fs *realSysFs) GetSystemUUID() (string, error) {
 	}
 }
 
-func (fs *realSysFs) IsCPUOnline(dir string) bool {
-	cpuPath := fmt.Sprintf("%s/online", dir)
-	content, err := ioutil.ReadFile(cpuPath)
+func (fs *realSysFs) IsCPUOnline(cpuPath string) bool {
+	return isCPUOnline(cpuPath)
+}
+
+// isCPUOnline verifies if a CPU in online. Takes path to /sys/devices/system/cpu/cpu*/ directory as a parameter.
+func isCPUOnline(cpuPath string) bool {
+	if isx86() && isZeroCPU(cpuPath) {
+		// It *is* possible to enable CPU hotplug for CPU0 but it requires setting a boot flag.
```

Comment:
I believe it is safe, but I can't promise it. Do you think we should check kernel boot params?

Reply:
is there any reason we need special logic for cpu0?

Reply:
or actually just looking at my 32 core ubuntu GCE VM:

```
root@ubuntu-big:~# find /sys/devices/system/cpu/ | grep "online"
/sys/devices/system/cpu/cpu21/online
/sys/devices/system/cpu/cpu11/online
/sys/devices/system/cpu/cpu9/online
/sys/devices/system/cpu/cpu7/online
/sys/devices/system/cpu/cpu28/online
/sys/devices/system/cpu/cpu18/online
/sys/devices/system/cpu/cpu5/online
/sys/devices/system/cpu/cpu26/online
/sys/devices/system/cpu/cpu16/online
/sys/devices/system/cpu/cpu3/online
/sys/devices/system/cpu/cpu24/online
/sys/devices/system/cpu/online
/sys/devices/system/cpu/cpu14/online
/sys/devices/system/cpu/cpu1/online
/sys/devices/system/cpu/cpu22/online
/sys/devices/system/cpu/cpu12/online
/sys/devices/system/cpu/cpu30/online
/sys/devices/system/cpu/cpu20/online
/sys/devices/system/cpu/cpu10/online
/sys/devices/system/cpu/cpu8/online
/sys/devices/system/cpu/cpu29/online
/sys/devices/system/cpu/cpu19/online
/sys/devices/system/cpu/cpu6/online
/sys/devices/system/cpu/cpu27/online
/sys/devices/system/cpu/cpu17/online
/sys/devices/system/cpu/cpu4/online
/sys/devices/system/cpu/cpu25/online
/sys/devices/system/cpu/cpu15/online
/sys/devices/system/cpu/cpu2/online
/sys/devices/system/cpu/cpu23/online
/sys/devices/system/cpu/cpu13/online
/sys/devices/system/cpu/cpu31/online
```

cpu0 online doesn't exist. is it why this case is needed?

Reply:
In general you should not put cpu0 offline on x86 because it is used to boot the operating system. If it's unavailable then you might not be able to start your system. I am *extremely* surprised by what you shared about your VM: can you share your kernel boot params? Perhaps it's a case of disabled cpu0 and I bet you haven't altered your configuration.

Reply:
This is just a standard ubuntu GCE VM, created as follows

```
$ gcloud beta compute instances create ubuntu-big --machine-type=e2-standard-32 --image-family=ubuntu-2004-lts --image-project=ubuntu-os-cloud --boot-disk-size=256GB --boot-disk-type=pd-ssd --zone=us-central1-c
```

```
$ uname -a
Linux ubuntu-big 5.4.0-1029-gcp #31-Ubuntu SMP Wed Oct 21 19:38:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

Here's KConfig from
```
sudo cat /boot/config-$(uname -r)
```
https://gist.github.com/bobbypage/3e65f6a6ad649f082760b9674bd02c5b